File: backend/app/api/routes/admin/cost_dashboard.py
"""Cost Admin Dashboard API endpoints."""

from fastapi import APIRouter, Depends, Query, Request
from app.core.rate_limiter import limiter
from app.core.logging_config import get_logger
from app.models.user import User

from .cost_auth import (
    require_cost_read_permission,
    require_per_user_cost_access,
    require_top_spenders_permission,
)
from .cost_schemas import (
    CostBreakdownResponse,
    CostComparisonResponse,
    CostOverviewResponse,
    CostQueryParams,
    FinancialStatementResponse,
    TimelineDataPoint,
    TopSpendersResponse,
)

from app.services.admin import cost_dashboard_service as service

router = APIRouter(prefix="/admin/costs", tags=["admin-costs"])
logger = get_logger(__name__)


@router.get("/overview")
@limiter.limit("60/hour")
async def get_cost_overview(
    request: Request,
    current_user: User = Depends(require_cost_read_permission),
) -> CostOverviewResponse:
    """Get current P&L summary and key metrics."""
    return await service.get_overview()


@router.get("/timeline")
@limiter.limit("30/hour")
async def get_cost_timeline(
    request: Request,
    params: CostQueryParams = Depends(),
    current_user: User = Depends(require_cost_read_permission),
) -> list[TimelineDataPoint]:
    """Get cost trends over time period."""
    return await service.get_timeline(params)


@router.get("/breakdown")
@limiter.limit("30/hour")
async def get_cost_breakdown(
    request: Request,
    params: CostQueryParams = Depends(),
    current_user: User = Depends(require_cost_read_permission),
) -> CostBreakdownResponse:
    """Get detailed cost breakdown by category."""
    return await service.get_breakdown(params)


@router.get("/balance-sheet")
@limiter.limit("20/hour")
async def get_balance_sheet(
    request: Request,
    params: CostQueryParams = Depends(),
    current_user: User = Depends(require_cost_read_permission),
) -> FinancialStatementResponse:
    """Get P&L statement (balance sheet)."""
    return await service.get_balance_sheet(params)


@router.get("/per-minute")
@limiter.limit("20/hour")
async def get_cost_per_minute(
    request: Request,
    params: CostQueryParams = Depends(),
    current_user: User = Depends(require_cost_read_permission),
) -> dict:
    """Get cost per minute metrics."""
    return await service.get_cost_per_minute(params)


@router.get("/users/top-spenders")
@limiter.limit("3/hour")
async def get_top_spenders(
    request: Request,
    period: str = Query("monthly", pattern="^(weekly|monthly|quarterly|yearly)$"),
    limit: int = Query(20, ge=1, le=100),
    current_user: User = Depends(require_top_spenders_permission),
) -> TopSpendersResponse:
    """
    Get top spenders ranking (PII redacted).

    Only accessible to SUPER_ADMIN.
    User IDs are hashed, amounts are rounded to ranges.
    """
    return await service.get_top_spenders(period, limit)


@router.get("/comparison")
@limiter.limit("15/hour")
async def get_cost_comparison(
    request: Request,
    params: CostQueryParams = Depends(),
    current_user: User = Depends(require_cost_read_permission),
) -> CostComparisonResponse:
    """Compare permanent vs transient costs."""
    return await service.get_comparison(params)


@router.get("/users/{user_id}/breakdown")
@limiter.limit("10/hour")
async def get_user_cost_breakdown(
    user_id: str,
    request: Request,
    params: CostQueryParams = Depends(),
    auth: tuple = Depends(require_per_user_cost_access),
) -> dict:
    """Get per-user cost breakdown."""
    return await service.get_user_breakdown(user_id, params)

File: web/src/services/adminApi/costDashboard.ts
// Cost Dashboard API service integration

import api from "../api";

interface CostOverviewParams {
  scope: "system_wide" | "per_user";
  userId?: string;
}

interface CostTimelineParams {
  scope: "system_wide" | "per_user";
  userId?: string;
  granularity?: "hourly" | "daily" | "monthly";
  startDate?: string;
  endDate?: string;
}

export const costDashboardService = {
  async getOverview(scope: "system_wide" | "per_user", userId?: string) {
    const query = new URLSearchParams({
      scope,
      ...(userId && { user_id: userId }),
    });

    return api.get(`/admin/costs/overview?${query.toString()}`);
  },

  async getTimeline(params: {
    scope: "system_wide" | "per_user";
    granularity?: "hourly" | "daily" | "monthly";
    startDate: string;
    endDate: string;
    userId?: string;
  }) {
    const query = new URLSearchParams({
      scope: params.scope,
      ...(params.granularity && { granularity: params.granularity }),
      start_date: params.startDate,
      end_date: params.endDate,
      ...(params.userId && { user_id: params.userId }),
    });

    return api.get(`/admin/costs/timeline?${query.toString()}`);
  },

  async getBreakdown(params: {
    period?: "month" | "year";
    startDate?: string;
    endDate?: string;
    scope: "system_wide" | "per_user";
    userId?: string;
  }) {
    const query = new URLSearchParams({
      ...(params.period && { period: params.period }),
      ...(params.startDate && { start_date: params.startDate }),
      ...(params.endDate && { end_date: params.endDate }),
      scope: params.scope,
      ...(params.userId && { user_id: params.userId }),
    });

    return api.get(`/admin/costs/breakdown?${query.toString()}`);
  },

  async getBalanceSheet(params: {
    period?: "month" | "year";
    startDate?: string;
    endDate?: string;
    scope: "system_wide" | "per_user";
    userId?: string;
  }) {
    const query = new URLSearchParams({
      ...(params.period && { period: params.period }),
      ...(params.startDate && { start_date: params.startDate }),
      ...(params.endDate && { end_date: params.endDate }),
      scope: params.scope,
      ...(params.userId && { user_id: params.userId }),
    });

    return api.get(`/admin/costs/balance-sheet?${query.toString()}`);
  },

  async getPerMinute(params: {
    period: "today" | "month";
    startDate?: string;
    endDate?: string;
    scope: "system_wide" | "per_user";
    userId?: string;
  }) {
    const query = new URLSearchParams({
      ...(params.period && { period: params.period }),
      ...(params.startDate && { start_date: params.startDate }),
      ...(params.endDate && { end_date: params.endDate }),
      scope: params.scope,
      ...(params.userId && { user_id: params.userId }),
    });

    return api.get(`/admin/costs/per-minute?${query.toString()}`);
  },

  async getTopSpenders(params: { period?: string; limit?: number }) {
    const query = new URLSearchParams({
      ...(params.period && { period: params.period }),
      ...(params.limit && { limit: params.limit.toString() }),
    });

    return api.get(`/admin/costs/users/top-spenders?${query.toString()}`);
  },

  async getComparison(params: {
    period: string;
    startDate?: string;
    endDate?: string;
    scope: "system_wide" | "per_user";
    userId?: string;
  }) {
    const query = new URLSearchParams({
      ...(params.period && { period: params.period }),
      ...(params.startDate && { start_date: params.startDate }),
      ...(params.endDate && { end_date: params.endDate }),
      scope: params.scope,
      ...(params.userId && { user_id: params.userId }),
    });

    return api.get(`/admin/costs/comparison?${query.toString()}`);
  },

  async getUserBreakdown(userId: string, dates?: { startDate: string; endDate: string }) {
    return api.get(`/admin/costs/users/${encodeURIComponent(userId)}/breakdown`, dates ? { params: { start_date: dates.startDate, end_date: dates.endDate } } : undefined);
  },
};
